python/lsst/scarlet/lite/io.py

```diff
@@ -0,0 +1,598 @@
+
+    origin: tuple[int, int]
+    peak: tuple[float, float]
+    model: np.ndarray
+
+    @property
+    def shape(self):
+        return self.model.shape[-2:]
+
+    def as_dict(self) -> dict:
+        """Return the object encoded into a dict for JSON serialization
+
+        Returns
+        -------
+        result:
+            The object encoded as a JSON compatible dict
+        """
+        return {
+            "origin": self.origin,
+            "shape": self.model.shape,
+            "peak": self.peak,
+            "model": tuple(self.model.flatten().astype(float)),
+        }
+
+    @classmethod
+    def from_dict(cls, data: dict, dtype: DTypeLike = np.float32) -> ScarletComponentData:
+        """Reconstruct `ScarletComponentData` from JSON compatible dict
+
+        Parameters
+        ----------
+        data:
+            Dictionary representation of the object
+        dtype:
+            Datatype of the resulting model.
+
+        Returns
+        -------
+        result:
+            The reconstructed object
+        """
+        data_shallow_copy = dict(data)
+        data_shallow_copy["origin"] = tuple(data["origin"])
+        shape = tuple(data_shallow_copy.pop("shape"))
+        data_shallow_copy["model"] = np.array(data["model"]).reshape(shape).astype(dtype)
+        return cls(**data_shallow_copy)
+
+
+@dataclass
+class ScarletFactorizedComponentData:
+    """Data for a factorized component
+
+    Attributes
+    ----------
+    origin:
+        The lower bound of the component's bounding box.
+    peak:
+        The ``(y, x)`` peak of the component.
+    spectrum:
+        The SED of the component.
+    morph:
+        The 2D morphology of the component.
+    """
+
+    origin: tuple[int, int]
+    peak: tuple[float, float]
+    spectrum: np.ndarray
+    morph: np.ndarray
+
+    @property
+    def shape(self):
+        return self.morph.shape
+
+    def as_dict(self) -> dict:
+        """Return the object encoded into a dict for JSON serialization
+
+        Returns
+        -------
+        result:
+            The object encoded as a JSON compatible dict
+        """
+        return {
+            "origin": tuple(int(o) for o in self.origin),
+            "shape": tuple(int(s) for s in self.morph.shape),
+            "peak": tuple(int(p) for p in self.peak),
+            "spectrum": tuple(self.spectrum.astype(float)),
+            "morph": tuple(self.morph.flatten().astype(float)),
+        }
+
+    @classmethod
+    def from_dict(cls, data: dict, dtype: DTypeLike = np.float32) -> ScarletFactorizedComponentData:
+        """Reconstruct `ScarletFactorizedComponentData` from JSON compatible
+        dict.
+
+        Parameters
+        ----------
+        data:
+            Dictionary representation of the object
+        dtype:
+            Datatype of the resulting model.
+
+        Returns
+        -------
+        result:
+            The reconstructed object
+        """
+        data_shallow_copy = dict(data)
```

Comment:
Yeah, I was just thinking this was more future proof. But I could also create a new dict if that makes you happier.

Reply:
I think it would be clearer to just create the dict and manually add new keys if needed, but I leave it to you since you have a better idea of what might change in the future.